File: java/src/main/java/io/hron/display/Display.java
package io.hron.display;

import io.hron.ast.*;
import java.util.List;
import java.util.stream.Collectors;

/** Renders schedule data as canonical strings. */
public final class Display {
  private Display() {}

  /**
   * Renders a schedule data as a canonical string.
   *
   * @param data the schedule data to render
   * @return the canonical string representation
   */
  public static String render(ScheduleData data) {
    StringBuilder sb = new StringBuilder();

    sb.append(renderExpr(data.expr()));

    if (!data.except().isEmpty()) {
      sb.append(" except ");
      sb.append(renderExceptions(data.except()));
    }

    if (data.until() != null) {
      sb.append(" until ");
      sb.append(renderUntil(data.until()));
    }

    if (data.anchor() != null && !data.anchor().isEmpty()) {
      sb.append(" starting ");
      sb.append(data.anchor());
    }

    if (!data.during().isEmpty()) {
      sb.append(" during ");
      sb.append(renderMonthList(data.during()));
    }

    if (data.timezone() != null && !data.timezone().isEmpty()) {
      sb.append(" in ");
      sb.append(data.timezone());
    }

    return sb.toString();
  }

  private static String renderExpr(ScheduleExpr expr) {
    return switch (expr) {
      case DayRepeat dr -> renderDayRepeat(dr);
      case IntervalRepeat ir -> renderIntervalRepeat(ir);
      case WeekRepeat wr -> renderWeekRepeat(wr);
      case MonthRepeat mr -> renderMonthRepeat(mr);
      case OrdinalRepeat or -> renderOrdinalRepeat(or);
      case SingleDate sd -> renderSingleDate(sd);
      case YearRepeat yr -> renderYearRepeat(yr);
    };
  }

  private static String renderDayRepeat(DayRepeat dr) {
    if (dr.interval() > 1) {
      return String.format("every %d days at %s", dr.interval(), formatTimeList(dr.times()));
    }
    return String.format("every %s at %s", renderDayFilter(dr.days()), formatTimeList(dr.times()));
  }

  private static String renderIntervalRepeat(IntervalRepeat ir) {
    StringBuilder sb = new StringBuilder();
    sb.append(
        String.format(
            "every %d %s from %s to %s",
            ir.interval(), ir.unit().display(ir.interval()), ir.fromTime(), ir.toTime()));
    if (ir.dayFilter() != null) {
      sb.append(" on ");
      sb.append(renderDayFilter(ir.dayFilter()));
    }
    return sb.toString();
  }

  private static String renderWeekRepeat(WeekRepeat wr) {
    return String.format(
        "every %d weeks on %s at %s",
        wr.interval(), formatDayList(wr.weekDays()), formatTimeList(wr.times()));
  }

  private static String renderMonthRepeat(MonthRepeat mr) {
    String targetStr = renderMonthTarget(mr.target());
    if (mr.interval() > 1) {
      return String.format(
          "every %d months on the %s at %s", mr.interval(), targetStr, formatTimeList(mr.times()));
    }
    return String.format("every month on the %s at %s", targetStr, formatTimeList(mr.times()));
  }

  private static String renderOrdinalRepeat(OrdinalRepeat or) {
    if (or.interval() > 1) {
      return String.format(
          "%s %s of every %d months at %s",
          or.ordinal(), or.weekday(), or.interval(), formatTimeList(or.times()));
    }
    return String.format(
        "%s %s of every month at %s", or.ordinal(), or.weekday(), formatTimeList(or.times()));
  }

  private static String renderSingleDate(SingleDate sd) {
    return String.format("on %s at %s", renderDateSpec(sd.dateSpec()), formatTimeList(sd.times()));
  }

  private static String renderYearRepeat(YearRepeat yr) {
    String targetStr = renderYearTarget(yr.target());
    if (yr.interval() > 1) {
      return String.format(
          "every %d years on %s at %s", yr.interval(), targetStr, formatTimeList(yr.times()));
    }
    return String.format("every year on %s at %s", targetStr, formatTimeList(yr.times()));
  }

  private static String renderDayFilter(DayFilter f) {
    return switch (f.kind()) {
      case EVERY -> "day";
      case WEEKDAY -> "weekday";
      case WEEKEND -> "weekend";
      case DAYS -> formatDayList(f.days());
    };
  }

  private static String renderMonthTarget(MonthTarget target) {
    return switch (target.kind()) {
      case LAST_DAY -> "last day";
      case LAST_WEEKDAY -> "last weekday";
      case DAYS -> formatOrdinalDaySpecs(target.specs());
    };
  }

  private static String renderYearTarget(YearTarget target) {
    return switch (target.kind()) {
      case DATE -> String.format("%s %d", target.month(), target.day());
      case ORDINAL_WEEKDAY ->
          String.format("the %s %s of %s", target.ordinal(), target.weekday(), target.month());
      case DAY_OF_MONTH ->
          String.format("the %s of %s", ordinalNumber(target.day()), target.month());
      case LAST_WEEKDAY -> String.format("the last weekday of %s", target.month());
    };
  }

  private static String renderDateSpec(DateSpec spec) {
    return switch (spec.kind()) {
      case NAMED -> String.format("%s %d", spec.month(), spec.day());
      case ISO -> spec.date();
    };
  }

  private static String renderExceptions(List<ExceptionSpec> exceptions) {
    return exceptions.stream().map(Display::renderExceptionSpec).collect(Collectors.joining(", "));
  }

  private static String renderExceptionSpec(ExceptionSpec exc) {
    return switch (exc.kind()) {
      case NAMED -> String.format("%s %d", exc.month(), exc.day());
      case ISO -> exc.date();
    };
  }

  private static String renderUntil(UntilSpec until) {
    return switch (until.kind()) {
      case ISO -> until.date();
      case NAMED -> String.format("%s %d", until.month(), until.day());
    };
  }

  private static String renderMonthList(List<MonthName> months) {
    return months.stream().map(MonthName::toString).collect(Collectors.joining(", "));
  }

  private static String formatTimeList(List<TimeOfDay> times) {
    return times.stream().map(TimeOfDay::toString).collect(Collectors.joining(", "));
  }

  private static String formatDayList(List<Weekday> days) {
    return days.stream().map(Weekday::toString).collect(Collectors.joining(", "));
  }

  private static String formatOrdinalDaySpecs(List<DayOfMonthSpec> specs) {
    return specs.stream().map(Display::formatDayOfMonthSpec).collect(Collectors.joining(", "));
  }

  private static String formatDayOfMonthSpec(DayOfMonthSpec spec) {
    return switch (spec.kind()) {
      case SINGLE -> ordinalNumber(spec.day());
      case RANGE ->
          String.format("%s to %s", ordinalNumber(spec.start()), ordinalNumber(spec.end()));
    };
  }

  private static String ordinalNumber(int n) {
    return n + ordinalSuffix(n);
  }

  private static String ordinalSuffix(int n) {
    int mod100 = n % 100;
    if (mod100 >= 11 && mod100 <= 13) {
      return "th";
    }
    return switch (n % 10) {
      case 1 -> "st";
      case 2 -> "nd";
      case 3 -> "rd";
      default -> "th";
    };
  }
}

File: csharp/Hron/Display/Display.cs
using System.Text;
using Hron.Ast;

namespace Hron.Display;

/// <summary>
/// Renders schedule data as canonical strings.
/// </summary>
public static class Display
{
    /// <summary>
    /// Renders a schedule data as a canonical string.
    /// </summary>
    public static string Render(ScheduleData data)
    {
        var sb = new StringBuilder();

        sb.Append(RenderExpr(data.Expr));

        if (data.Except.Count > 0)
        {
            sb.Append(" except ");
            sb.Append(RenderExceptions(data.Except));
        }

        if (data.Until is not null)
        {
            sb.Append(" until ");
            sb.Append(RenderUntil(data.Until));
        }

        if (!string.IsNullOrEmpty(data.Anchor))
        {
            sb.Append(" starting ");
            sb.Append(data.Anchor);
        }

        if (data.During.Count > 0)
        {
            sb.Append(" during ");
            sb.Append(RenderMonthList(data.During));
        }

        if (!string.IsNullOrEmpty(data.Timezone))
        {
            sb.Append(" in ");
            sb.Append(data.Timezone);
        }

        return sb.ToString();
    }

    private static string RenderExpr(IScheduleExpr expr) => expr switch
    {
        DayRepeat dr => RenderDayRepeat(dr),
        IntervalRepeat ir => RenderIntervalRepeat(ir),
        WeekRepeat wr => RenderWeekRepeat(wr),
        MonthRepeat mr => RenderMonthRepeat(mr),
        SingleDate sd => RenderSingleDate(sd),
        YearRepeat yr => RenderYearRepeat(yr),
        _ => throw new ArgumentException($"Unknown expression type: {expr.GetType()}", nameof(expr))
    };

    private static string RenderDayRepeat(DayRepeat dr)
    {
        if (dr.Interval > 1)
        {
            return $"every {dr.Interval} days at {FormatTimeList(dr.Times)}";
        }
        return $"every {RenderDayFilter(dr.Days)} at {FormatTimeList(dr.Times)}";
    }

    private static string RenderIntervalRepeat(IntervalRepeat ir)
    {
        var sb = new StringBuilder();
        sb.Append($"every {ir.Interval} {ir.Unit.Display(ir.Interval)} from {ir.FromTime} to {ir.ToTime}");
        if (ir.DayFilter is not null)
        {
            sb.Append(" on ");
            sb.Append(RenderDayFilter(ir.DayFilter));
        }
        return sb.ToString();
    }

    private static string RenderWeekRepeat(WeekRepeat wr)
        => $"every {wr.Interval} weeks on {FormatDayList(wr.WeekDays)} at {FormatTimeList(wr.Times)}";

    private static string RenderMonthRepeat(MonthRepeat mr)
    {
        var targetStr = RenderMonthTarget(mr.Target);
        if (mr.Interval > 1)
        {
            return $"every {mr.Interval} months on the {targetStr} at {FormatTimeList(mr.Times)}";
        }
        return $"every month on the {targetStr} at {FormatTimeList(mr.Times)}";
    }

    private static string RenderSingleDate(SingleDate sd)
        => $"on {RenderDateSpec(sd.DateSpec)} at {FormatTimeList(sd.Times)}";

    private static string RenderYearRepeat(YearRepeat yr)
    {
        var targetStr = RenderYearTarget(yr.Target);
        if (yr.Interval > 1)
        {
            return $"every {yr.Interval} years on {targetStr} at {FormatTimeList(yr.Times)}";
        }
        return $"every year on {targetStr} at {FormatTimeList(yr.Times)}";
    }

    private static string RenderDayFilter(DayFilter f) => f.Kind switch
    {
        DayFilterKind.Every => "day",
        DayFilterKind.Weekday => "weekday",
        DayFilterKind.Weekend => "weekend",
        DayFilterKind.Days => FormatDayList(f.Days),
        _ => throw new ArgumentOutOfRangeException()
    };

    private static string RenderMonthTarget(MonthTarget target) => target.Kind switch
    {
        MonthTargetKind.LastDay => "last day",
        MonthTargetKind.LastWeekday => "last weekday",
        MonthTargetKind.Days => FormatOrdinalDaySpecs(target.Specs),
        MonthTargetKind.NearestWeekday => RenderNearestWeekday(target),
        MonthTargetKind.OrdinalWeekday => $"{target.OrdinalValue!.Value.ToDisplayString()} {target.WeekdayValue!.Value.ToDisplayString()}",
        _ => throw new ArgumentOutOfRangeException()
    };

    private static string RenderNearestWeekday(MonthTarget target)
    {
        var prefix = target.NearestWeekdayDirection switch
        {
            NearestDirection.Next => "next ",
            NearestDirection.Previous => "previous ",
            _ => ""
        };
        return $"{prefix}nearest weekday to {OrdinalNumber(target.NearestWeekdayDay)}";
    }

    private static string RenderYearTarget(YearTarget target) => target.Kind switch
    {
        YearTargetKind.Date => $"{target.Month.ToDisplayString()} {target.Day}",
        YearTargetKind.OrdinalWeekday => $"the {target.Ordinal!.Value.ToDisplayString()} {target.WeekdayValue!.Value.ToDisplayString()} of {target.Month.ToDisplayString()}",
        YearTargetKind.DayOfMonth => $"the {OrdinalNumber(target.Day)} of {target.Month.ToDisplayString()}",
        YearTargetKind.LastWeekday => $"the last weekday of {target.Month.ToDisplayString()}",
        _ => throw new ArgumentOutOfRangeException()
    };

    private static string RenderDateSpec(DateSpec spec) => spec.Kind switch
    {
        DateSpecKind.Named => $"{spec.Month!.Value.ToDisplayString()} {spec.Day}",
        DateSpecKind.Iso => spec.Date!,
        _ => throw new ArgumentOutOfRangeException()
    };

    private static string RenderExceptions(IReadOnlyList<ExceptionSpec> exceptions)
        => string.Join(", ", exceptions.Select(RenderExceptionSpec));

    private static string RenderExceptionSpec(ExceptionSpec exc) => exc.Kind switch
    {
        ExceptionSpecKind.Named => $"{exc.Month!.Value.ToDisplayString()} {exc.Day}",
        ExceptionSpecKind.Iso => exc.Date!,
        _ => throw new ArgumentOutOfRangeException()
    };

    private static string RenderUntil(UntilSpec until) => until.Kind switch
    {
        UntilSpecKind.Iso => until.Date!,
        UntilSpecKind.Named => $"{until.Month!.Value.ToDisplayString()} {until.Day}",
        _ => throw new ArgumentOutOfRangeException()
    };

    private static string RenderMonthList(IReadOnlyList<MonthName> months)
        => string.Join(", ", months.Select(m => m.ToDisplayString()));

    private static string FormatTimeList(IReadOnlyList<TimeOfDay> times)
        => string.Join(", ", times.Select(t => t.ToString()));

    private static string FormatDayList(IReadOnlyList<Weekday> days)
        => string.Join(", ", days.Select(d => d.ToDisplayString()));

    private static string FormatOrdinalDaySpecs(IReadOnlyList<DayOfMonthSpec> specs)
        => string.Join(", ", specs.Select(FormatDayOfMonthSpec));

    private static string FormatDayOfMonthSpec(DayOfMonthSpec spec) => spec.Kind switch
    {
        DayOfMonthSpecKind.Single => OrdinalNumber(spec.Day),
        DayOfMonthSpecKind.Range => $"{OrdinalNumber(spec.Start)} to {OrdinalNumber(spec.End)}",
        _ => throw new ArgumentOutOfRangeException()
    };

    private static string OrdinalNumber(int n) => n + OrdinalSuffix(n);

    private static string OrdinalSuffix(int n)
    {
        var mod100 = n % 100;
        if (mod100 is >= 11 and <= 13)
        {
            return "th";
        }
        return (n % 10) switch
        {
            1 => "st",
            2 => "nd",
            3 => "rd",
            _ => "th"
        };
    }
}
